### ISSUE TITLE
More lazy container operations

### ISSUE THREAD

**Issue by richardstartin** [RoaringBitmap/RoaringBitmap#212]
I think it would be faster to evaluate circuits containing `and`s, `andnot`s and `xor`s if these could be done laziliy, in order to skip any bitmap cardinality calculations of intermediate results. A typical use case would be implementing a container-wise implementation of the logic [here](https://github.com/richardstartin/isometric/blob/master/index/src/main/java/com/openkappa/isometric/index/bitslice/RangeEncodedOptBitSliceIndex.java#L42).

This is implemented for `or` already. Was this ever attempted or measured with the other operations?

**Comment by lemire** [comment]
I think we optimized unions (``or``)  because that's a common use case, and it is a case where we can expect huge gains. As discussed elsewhere, intersections are already fast in many instances, so the expected gains are less.

Of course, if you are doing a circuit evaluation where you mix and match various operations (and, or, xor, andnot), then it is a different story.

**Comment by richardstartin** [comment]
If a PR is welcome for this I will implement these methods soon.

**Comment by lemire** [comment]
Welcome!

**Comment by richardstartin** [comment]
Is this megamorphism evasion?

```java

**Comment by lemire** [comment]
I don't think so. We don't just cast the types... look carefully. It could be implemented more elegantly, but ultimately, we have a 3x3 matrix mapping containers to functions... so it is going to be equivalent to this.
